### ISSUE TITLE
How can I use Scrcpy with Python?

### ISSUE THREAD

**Issue by CZSZCX** [Genymobile/scrcpy#5443]
I want to use scrcpy in a Python file or command prompt, but I don't need to display windows in real-time. I just want its "simulate physical keyboard and mouse" feature. How can I do it?

**Comment by rom1v** [comment]
Use the same [protocol](https://github.com/Genymobile/scrcpy/blob/master/doc/develop.md#protocol) as the scrcpy client.

**Comment by CZSZCX** [comment]
So far, it seems that Scrcpy does not have the function of directly using cmd to send keyboard and mouse information. What's more, due to the high latency of Scrcpy's screen synchronization function, I chose to use a capture card to achieve a higher refresh rate and lower latency. But when I was using Scrcpy, I also found that the image transmission of the capture card was interrupted and had to be unplugged again to capture images normally.

**Comment by rom1v** [comment]
> So far, it seems that Scrcpy does not have the function of directly using cmd to send keyboard and mouse information.

This is correct: scrcpy does not expose an interface to send commands "programmatically".

> What's more, due to the high latency of Scrcpy's screen synchronization function

What do you call "high latency"? If you have more than a few tens of milliseconds, there's a problem (but for some use cases, it can be called "high latency").

> Therefore, I currently do not want Scrcpy to create windows, but at the same time, I would like to use hid/aoa keyboard and mouse simulation

Use `scrcpy --no-video --no-audio -KM` to just get controls (but you'll have a window to capture mouse/keyboard events).

**Comment by CZSZCX** [comment]
https://github.com/user-attachments/assets/ba22f778-f485-4018-a706-e4129970ff4a

**Comment by CZSZCX** [comment]
But I want to achieve simultaneous control of multiple devices and automatically perform simple operations in some cases (without using adb)

**Comment by rom1v** [comment]
> After pausing for a while, the delay will become "0" again when the next operation starts

Is it over wifi or over USB? 

What's the mirroring resolution? Does it also happen with a lower definition (`scrcpy -m1024`)?

Do you mirror a real device, or an emulator?

What's the software in the bottom-right?

**Comment by CZSZCX** [comment]
> > 暂停一段时间后，下次操作开始时，延迟会再次变为 “0”
> 
> 是通过 wifi 还是通过 USB？
> 
> 镜像分辨率是多少？它是否也发生在较低的定义 （） 中？`scrcpy -m1024`
> 
> 您镜像的是真实设备还是仿真器？
> 
> 右下角的软件是什么？

Through USB, the resolution is 1080p, for real devices. The software in the bottom right corner is a P2P remote control (with a latency of around 10ms and stable because it is a local area network). My Scrcpy test runs on a desktop computer and now uses a laptop for remote control.

I didn't pay attention to how to reduce the resolution of scrcpy (because I didn't originally intend to use scrcpy for image transfer function)

**Comment by CZSZCX** [comment]
> > > 暂停一段时间后，下次操作开始时，延迟会再次变为 “0”
> > 
> > 
> > 是通过 wifi 还是通过 USB？
> > 镜像分辨率是多少？它是否也发生在较低的定义 （） 中？`scrcpy -m1024`
> > 您镜像的是真实设备还是仿真器？
> > 右下角的软件是什么？
> 
> Through USB, the resolution is 1080p, real device. The software in the bottom right corner is a P2P remote control (with a latency of around 10ms and stable because it is a local area network). My Scrcpy test runs on a desktop computer and now uses a laptop for remote control.
> 
> I didn't pay attention to how to reduce the resolution of scrcpy (because I didn't originally intend to use scrcpy for image transfer function)

**Comment by rom1v** [comment]
> Through USB, the resolution is 1080p, real device.

It should work perfectly then. I don't understand your ultra-high latency.

I'm not sure to understand your remote control setup. Is there any remote control involved in your scrcpy setup? Or your phone is directly plugged to the computer where you run scrcpy?

**Comment by CZSZCX** [comment]
My Android device is directly plugged into the USB port of the computer.

**Comment by CZSZCX** [comment]
> 这是正确的： scrcpy 没有公开一个接口来 “以编程方式” 发送命令。

Is it possible to add this feature in future updates? 


> 使用与 scrcpy 客户端相同的[协议](https://github.com/Genymobile/scrcpy/blob/master/doc/develop.md#protocol)。

Or I can only try using the same protocol as the scrcpy client to meet my needs (but I don't think I can implement this feature, which is difficult for me)

**Comment by CZSZCX** [comment]
> This is correct: scrcpy does not expose an interface to send commands "programmatically".

Is it possible to add this feature in future updates?



> Use the same [protocol](https://github.com/Genymobile/scrcpy/blob/master/doc/develop.md#protocol) as the scrcpy client.

Or I can only try using the same protocol as the scrcpy client to meet my needs (but I don't think I can implement this feature, which is difficult for me)

**Comment by rom1v** [comment]
If you record (`scrcpy --record=file.mp4`), is the recorded file laggy?

Try with another encoder: https://github.com/Genymobile/scrcpy/blob/master/doc/video.md#encoder

**Comment by CZSZCX** [comment]
> If you record (`scrcpy --record=file.mp4`), is the recorded file laggy?
> 
> Try with another encoder: https://github.com/Genymobile/scrcpy/blob/master/doc/video.md#encoder

[server] INFO: List of video encoders:
    --video-codec=h264 --video-encoder='c2.android.avc.encoder'
    --video-codec=h264 --video-encoder='OMX.google.h264.encoder'
    --video-codec=h265 --video-encoder='c2.android.hevc.encoder'
    --video-codec=av1 --video-encoder='c2.android.av1.encoder'
[server] INFO: List of audio encoders:
    --audio-codec=opus --audio-encoder='c2.android.opus.encoder'
    --audio-codec=aac --audio-encoder='c2.android.aac.encoder'
    --audio-codec=aac --audio-encoder='OMX.google.aac.encoder'
    --audio-codec=flac --audio-encoder='c2.android.flac.encoder'
    --audio-codec=flac --audio-encoder='OMX.google.flac.encoder'

**Comment by CZSZCX** [comment]
> Use the same [protocol](https://github.com/Genymobile/scrcpy/blob/master/doc/develop.md#protocol) as the scrcpy client.

Uh, I'll try it later. So, regarding my question, is this the only method currently available?

**Comment by CZSZCX** [comment]
Scrcpy seems to be the only project I have found so far that can send keyboard and mouse events without using adb

**Comment by CZSZCX** [comment]
Scrcpy seems to be the only project I have found that can send keyboard and mouse events without using adb. I have prepared an alternative solution for adb for my project, but the previous implementation was not very effective. I only achieved the same operation on Android devices after completing the computer operation. (Cannot send to Android device before operation is completed, and cannot implement zoom function)
Therefore, after seeing Scrcpy, I considered whether it was possible to use its function of sending keyboard and mouse signals. However, based on the current situation, it seems difficult for me because my knowledge of hardware communication protocols is poor.

**Comment by CZSZCX** [comment]
Sorry, I clicked the wrong one.

**Comment by me2sy** [comment]
> I want to use scrcpy in a Python file or command prompt, but I don't need to display windows in real-time. I just want its "simulate physical keyboard and mouse" feature. How can I do it?

如果必须使用python程序接入的话，可以看看 [MYScrcpy](https://github.com/me2sy/MYScrcpy)
在[core/control](https://github.com/me2sy/MYScrcpy/blob/main/src/myscrcpy/core/control.py)里有对 scrcpy protocol的解析与实现，希望能给你带来一点思路。

**Comment by CZSZCX** [comment]
> > I want to use scrcpy in a Python file or command prompt, but I don't need to display windows in real-time. I just want its "simulate physical keyboard and mouse" feature. How can I do it?
> 
> 如果必须使用python程序接入的话，可以看看 [MYScrcpy](https://github.com/me2sy/MYScrcpy) 在[core/control](https://github.com/me2sy/MYScrcpy/blob/main/src/myscrcpy/core/control.py)里有对 scrcpy protocol的解析与实现，希望能给你带来一点思路。

好的谢谢，我看看
